nemo/collections/asr/models/asr_model.py

```diff
@@ -136,9 +136,61 @@ def on_after_backward(self):
                 torch.distributed.all_reduce(valid_gradients, op=torch.distributed.ReduceOp.MIN)
 
             if valid_gradients < 1:
-                logging.warning(f'detected inf or nan values in gradients! Setting gradients to zero.')
+                logging.warning('detected inf or nan values in gradients! Setting gradients to zero.')
                 self.zero_grad()
 
+    def on_train_start(self):
+        super().on_train_start()
+
+        # dynamic freezing
+        # should fire only once, on the very first batch of training and never again
+        if not hasattr(self, '_freeze_cfg'):
+            if (
+                hasattr(self.cfg, 'freeze_updates')
+                and self.cfg.freeze_updates is not None
+                and self.cfg.freeze_updates.get('enabled', False)
+            ):
+                setattr(self, '_freeze_cfg', OmegaConf.to_container(self.cfg.freeze_updates))
+                self._freeze_cfg['is_frozen'] = {k: False for k in self._freeze_cfg['modules'].keys()}
+            else:
+                setattr(self, '_freeze_cfg', None)
+
+    def on_train_batch_start(self, batch: Any, batch_idx: int):
+        """
+        dynamic freezing
+        allows you to freeze certain model layers for the first N steps of training
+        """
+        super().on_train_batch_start(batch, batch_idx)
+
+        # dynamic freezing
+        if hasattr(self, '_freeze_cfg') and self._freeze_cfg is not None:
+            if self.training and hasattr(self, "trainer") and self.trainer is not None:
+                num_updates = self.trainer.global_step + 1
+
+                for ml, m_steps in self._freeze_cfg['modules'].items():
+                    # we could do hasattr check here, but it's too expensive for each step
+                    # consequently you'll throw an error if the module name doesn't exist
+                    # or was spelled wrong in the config.yaml
+                    if isinstance(m_steps, list):
```

Comment:
Have you tested this list feature?
I think the list which comes a conf has type of ListConfig, not list.

Reply:
Yes, I have tested it and it works. The type is actually a list, not ListConfig. If you look at line 153 in that same file, you will see we call `setattr(self, '_freeze_cfg', OmegaConf.to_container(self.cfg.freeze_updates))` which casts from OmegaConf types down to Pythonic types.